### PR TITLE
Improve HttpApp stability #824

### PR DESCRIPTION
**Pull request by jlprat**
Issue: #824 
Attempt to fix this issue, and make tests more reliable.
reduce calls, to save threads

### CODE REVIEW

**Comment by jlprat** [comment]
Test fails because it can't start an `ActorSystem` and the server within 2 seconds.

**Comment by jlprat** [comment]
Increased timeout. I couldn't test on Win machine though...

File: akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala [hunk 1]
```diff
@@ -22,48 +23,55 @@ import scala.runtime.BoxedUnit
 class HttpAppSpec extends AkkaSpec with RequestBuilding with Eventually {
   import system.dispatcher
 
-  "HttpApp" should {
+  "HttpApp  Java" should {
```
Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala]: one space

File: akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala [hunk 2]
```diff
@@ -22,48 +23,55 @@ import scala.runtime.BoxedUnit
...
 
     "start without ActorSystem" in {
-      val (_, host, port) = TestUtils.temporaryServerHostnameAndPort()
 
       val minimal = new MinimalHttpApp()
 
       val server = Future {
-        minimal.startServer(host, port, ServerSettings.create(ConfigFactory.load))
+        minimal.startServer("localhost", 0, ServerSettings.create(ConfigFactory.load))
```
Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala]: Seems weird still that this would be more stable... why was that?

Comment (jonas) [inline akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala]: As mentioned in #824 there is a potential race condition when probing a port using `temporaryServerHostnameAndPort `.

File: akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala [hunk 3]
```diff
@@ -22,48 +23,55 @@ import scala.runtime.BoxedUnit
...
-      // Requesting the server to shutdown
-      callAndVerify(host, port, "shutdown")
+      try {
+        // Requesting the server to shutdown
+        callAndVerify(binding.localAddress.getHostName, binding.localAddress.getPort, "shutdown")
 
-      Await.ready(server, Duration(1, TimeUnit.SECONDS))
-      server.isCompleted should ===(true)
-      system.whenTerminated.isCompleted should ===(false)
+        Await.ready(server, Duration(1, TimeUnit.SECONDS))
```
Comment (jonas) [inline akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala]: Given the amount of additional code to ensure proper shutdown would it make sense to also change tests to either use test fixtures/contexts or before/after hooks to handle this?

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/HttpAppSpec.scala [hunk 4]
```diff
@@ -50,143 +56,155 @@ class HttpAppSpec extends AkkaSpec with Directives with RequestBuilding with Eve
...
+        }
 
         override protected def postServerShutdown(attempt: Try[Done], system: ActorSystem): Unit = postShutdownCalled.set(true)
       }
 
       "shutting down" in {
-        val (_, host, port) = TestUtils.temporaryServerHostnameAndPort()
 
-        val sneaky = new SneakSever()
+        val sneaky = new SneakySever()
```
Comment (jonas) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/HttpAppSpec.scala]: s/Sever/Server/

File: akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala (hunk 2 above)
Comment (jlprat) [inline akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala]: I think now I know why it failed that time, and the reason behind the deadlock in win.
It's not port collision (the stack trace we saw was the expected from binding failure test), it is that the server in rare occasions can't start in 5 seconds (start the actor system and the server). This causes an assertion failure, which prevents the server to be correctly shutdown, which is waiting for a `Future` to complete.
I could go back to the random port utility as it's simpler

File: akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala (hunk 3 above)
Comment (jlprat) [inline akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala]: Will look into it

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/HttpAppSpec.scala [hunk 5]
```diff
@@ -42,23 +43,63 @@ class HttpAppSpec extends AkkaSpec with Directives with RequestBuilding with Eve
...
   }
 
-  "HttpApp" should {
+  class SneakyServer extends MinimalApp {
+
+    val postBindingCalled = new AtomicBoolean(false)
+    val postBindingFailureCalled = new AtomicBoolean(false)
+    val postShutdownCalled = new AtomicBoolean(false)
+
+    override protected def waitForShutdownSignal(system: ActorSystem)(implicit ec: ExecutionContext): Future[Done] = {
```
Comment (jonas) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/HttpAppSpec.scala]: This `def` could be removed since it is inherited.

Comment (jlprat) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/HttpAppSpec.scala]: True, will do in a sec

**Comment by jrudolph** [comment]
Thanks a lot, @jlprat.

**Pull request merged**